lib/SILOptimizer/Utils/OptimizerStatsUtils.cpp

```diff
@@ -0,0 +1,970 @@
+/// that changes in the statistics are only reported if the delta between the
+/// old and the new values are at least X%. The deltas are computed using the
+/// following formula:
+///  Delta = (NewValue - OldValue) / OldValue
+///
+/// Thresholds provide a simple way to perform a simple filtering of the
+/// collected statistics during the compilation. But if there is a need for a
+/// more complex analysis of collected data (e.g. aggregation by a pipeline
+/// stage or by the type of a transformation), it is often better to dump
+/// as much data as possible into a file using e.g. `-sil-stats-dump-all
+/// -sil-stats-modules -sil-stats-functions` and then e.g. use the helper scripts
+/// to store the collected data into a database and then perform complex queries
+/// on it. Many kinds of analysis can be then formulated pretty easily as SQL
+/// queries.
+///
+/// The output format is a set of CSV (comma separated values) lines. Each lines
+/// represents one counter or one counter change.
+///
+/// For counter updates it looks like this:
+///  Kind, CounterName, StageName, TransformName, TransformPassNumber,
+///  DeltaValue, OldCounterValue, NewCounterValue, Duration, Symbol
+///
+/// For counter updates it looks like this:
+///  Kind, CounterName, StageName, TransformName, TransformPassNumber,
+///  CounterValue, Duration, Symbol
+///
+/// where Kind is one of "function", "module", "function_history",
+///       CounterName is one of "block", "inst", "function", "memory",
+///       Symbol is e.g. the name of a function.
+///       StageName and TransformName are the names of the current optimizer
+///       pipeline stage and current transform.
+///       Duration is the duration of the transformation.
+//===----------------------------------------------------------------------===//
+
+#include <iostream>
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Process.h"
+#include "swift/SIL/SILValue.h"
+#include "swift/SIL/SILVisitor.h"
+#include "swift/SILOptimizer/Analysis/Analysis.h"
+#include "swift/SILOptimizer/PassManager/PassManager.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/OptimizerStatsUtils.h"
+using namespace swift;
+
+namespace {
+
+/// The total number of different kinds of SILInstructions.
+constexpr int SILInstructionsNum = int(ValueKind::Last_SILInstruction) + 1;
+
+/// A set of counters, one per SILInstruction kind.
+typedef SmallVector<int, SILInstructionsNum> InstructionCounts;
+
+/// Map a SILInstruction name to its SILInstructionKind.
+SILInstructionKind getSILInstructionKind(StringRef InstName) {
+#define INST(Id, Parent, TextualName, MemoryBehavior, ReleasingBehavior)       \
+  if (InstName == #TextualName)                                                \
+    return SILInstructionKind::Id;
+
+#include "swift/SIL/SILNodes.def"
+
+#ifdef NDEBUG
+  std::cerr
```

Comment:
Please do not use std::cerr. Please use:

```
llvm::errs()
```

Also, there is a typo:
```
insruction => instruction
```

Finally, why not just create an unreachable that always asserts irregardless of assertions. I have wanted that before.